docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+    const interfaceOption = interfaceSelect.find('option');
+    const value           = await getValue();
+
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this example, the client function obtains the value of a drop-down element. This value is then passed to the `t.expect` method. Since the client function is awaited when it's assigned to the variable, the value gets resolved instantly and never updates. The smart assertion query mechanism doesn't apply and the test fails.
+
+To solve the issue, pass the client function without the `await` keyword:
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with ClientFunction', async t => {
+    const getValue = ClientFunction(() => document.getElementById('preferred-interface').value)
+
+    const interfaceSelect = Selector('#preferred-interface');
+    const interfaceOption = interfaceSelect.find('option');
+    const value           =  getValue();
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this case, TestCafe applies the smart query mechanism and retries the assertion multiple times. This test responds to the changes that happen on the page and is more conclusive.
+
+## Use of Page Objects
+
+Use the Page Model in your tests. It helps you abstract the tested page into a single file that you can use to target page elements.
+
+This approach improves the flexibility of your tests - if the UI changes, change one file to update all the references. Define common operations as your model's methods. Tests are then more readable and less brittle because there is no duplicate code.
+
+The page model use is described in more detail in the [Page Model](../../guides/concepts/page-model.md) article.
+
+A code example is available in the [testcafe-examples](https://github.com/DevExpress/testcafe-examples/tree/master/examples/use-page-model) repository.
+
+## Use of Roles for Login
+
+Handle authentication during your tests with [User Roles](../../guides/advanced-guides/authentication.md#user-roles). See the [Authentication] article for more in-depth information.
+
+## File Structure
+
+Follow these guidelines to keep your test structure manageable and "clean":
+
+* If you use [page model](../../guides/concepts/page-model.md) for your tests (which is recommended), put all the page model files into one directory. If your application is divisible into logical chunks (components or subsystems), split up the associated page model objects into separate files.
+
+* Use the page model to store Selectors and compound actions that are common in your app. For instance, put all actions that are necessary to perform a login into one reusable function.
```

Comment:
I guess `login` here doesn't require an article, and I feel like *used frequently across your app* sounds better than `that are common in your app`.

Reply:
Roles are better suited for performing login. I've since rewritten:

> Use a page model to store Selectors and compound actions that are used frequently across your app. For instance, put all actions that are necessary to perform an action into one reusable function.